### PR TITLE
Add optional logging of text output to EvalOutputLogging

### PR DESCRIPTION
**Pull request by sjawhar**
Adds the ability to log text output using `EvalOutputLogging` in non-metrics/ICL use cases. This appears as a new `outputs` key in the logged dictionary, which is set to `stats.outpus` and de-tokenized when the model is a `HuggingFaceModel` and the dataset has a tokenizer.

### CODE REVIEW

File: llmfoundry/callbacks/eval_output_logging_callback.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ def eval_batch_end(self, state: State, logger: Logger) -> None:
                                       )
 
         # If batch mode is not generate, outputs will be logits
-        if state.batch['mode'] == 'generate':
+        if state.batch.get('mode') == 'generate':
```
Comment (sjawhar) [inline llmfoundry/callbacks/eval_output_logging_callback.py:51]: Should the default assumption be that it is indeed "generate"?

Comment (sjawhar) [reply llmfoundry/callbacks/eval_output_logging_callback.py:51]: Nope! Tested myself and no good.

**Comment by mvpatel2000** [comment]
@maxisawesome can you please take a look? 

CC: @dakinggg

**Comment by sjawhar** [comment]
image
YES!! It's finally working! The logging, that is. Not my fine-tuning, that's still garbage. But I can finally start debugging!

File: llmfoundry/callbacks/eval_output_logging_callback.py [hunk 2]
```diff
@@ -47,28 +47,31 @@ def eval_batch_end(self, state: State, logger: Logger) -> None:
...
             # Outputs are already detokenized
             logging_dict['outputs'] = state.outputs
+        elif isinstance(state.outputs, torch.Tensor):
+            # If batch mode is not generate, outputs will be logits
+            logging_dict['outputs'] = state.outputs.argmax(dim=-1)
 
         input_ids = state.batch['input_ids']
         logged_input = []
         assert state.dataloader is not None
+        dataset = state.dataloader.dataset  # pyright: ignore[reportGeneralTypeIssues]
```
Comment (maxisawesome) [inline llmfoundry/callbacks/eval_output_logging_callback.py:61]: This makes things much more legible but also doubles the dataset in memory 😔. Most eval datasets are small and it wouldn't matter much, but we have run into problems when people using surprisingly large eval datasets before, so I don't think this is a good idea in the end.

**Comment by maxisawesome** [comment]
Looks great other than the one comment I left! Thanks for fixing this for non-generative evals.

File: llmfoundry/callbacks/eval_output_logging_callback.py (hunk 2 above)
Comment (sjawhar) [reply llmfoundry/callbacks/eval_output_logging_callback.py:61]: Is this a gc issue? Isn't the variable already defined in the `for input_list in` loop?

Comment (mvpatel2000) [reply llmfoundry/callbacks/eval_output_logging_callback.py:61]: Discussed offline, resolving -- should not result in a copy since it stores a pointer.

**Comment by mvpatel2000** [comment]
@sjawhar if you can fix lint + unit tests (the CPU ones) that would be awesome! then we can merge

File: tests/callbacks/test_eval_output_logging_callback.py [hunk 3]
```diff
@@ -193,23 +205,27 @@ def test_eval_output_logging_lm(
...
     # We use the same data in each batch
-    assert json.loads(in_memory_logger.tables[f'lm_acc_step_0'])['data'] == [
+    assert logged_data['data'] == [
         [
             'The dog is',
             ' furry',
             ' furry',
             1,
             'InContextLearningLMAccuracy',
+            RegexMatcher(r' dog is furry(\[PAD\])+I'),
```
Comment (sjawhar) [inline tests/callbacks/test_eval_output_logging_callback.py:228]: So I'm not sure you'll like this approach :sweat_smile: 

I actually was surprised to find `output` already in the logged data, since it's not in `eval_output_logger`. Is that from the ICL metric/eval itself? It makes the new `outputs` field look superfluous (though also possibly not, since `outputs` is like a `raw_output`). This pattern matching here to pass the test also feels a bit unprincipled.

**Comment by sjawhar** [comment]
Anything else I can do to help get this merged?

**Comment by dakinggg** [comment]
@sjawhar sorry about that, will take a look this week!

File: tests/callbacks/test_eval_output_logging_callback.py (hunk 3 above)
Comment (dakinggg) [reply tests/callbacks/test_eval_output_logging_callback.py:228]: as for the outputs already being there, thats cause of https://github.com/mosaicml/llm-foundry/blob/21c9e0a58a68b2ddb7cc7bea40a11c9add85d2c9/llmfoundry/callbacks/eval_output_logging_callback.py#L53 right?

Comment (sjawhar) [reply tests/callbacks/test_eval_output_logging_callback.py:228]: Yes, I expected `outputs` (with an "s"), but not `output` (not with an "s"). Don't know where that's coming from.

File: llmfoundry/callbacks/eval_output_logging_callback.py [hunk 4]
```diff
@@ -47,28 +47,34 @@ def eval_batch_end(self, state: State, logger: Logger) -> None:
                                           state.metric_outputs,
                                       )
 
-        # If batch mode is not generate, outputs will be logits
-        if state.batch['mode'] == 'generate':
+        if state.batch.get('mode') == 'generate':
             # Outputs are already detokenized
             logging_dict['outputs'] = state.outputs
+        elif isinstance(state.outputs, torch.Tensor):
```
[the remark this answers is not in the document]
Comment (sjawhar) [reply llmfoundry/callbacks/eval_output_logging_callback.py:53]: I gave it a go

File: tests/callbacks/test_eval_output_logging_callback.py (hunk 3 above)
Comment (dakinggg) [reply tests/callbacks/test_eval_output_logging_callback.py:228]: oh i see what you mean. Yeah, the logging dict starts off as a deep copy of whatever the metric outputs were (https://github.com/mosaicml/llm-foundry/blob/88511f780c7437e89cdaf96a1771aac41c2a64a5/llmfoundry/callbacks/eval_output_logging_callback.py#L45-L48)

File: llmfoundry/callbacks/eval_output_logging_callback.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dakinggg) [reply llmfoundry/callbacks/eval_output_logging_callback.py:53]: possibly forgot to push this change?

**Review by dakinggg — APPROVED** [review]
LGTM, please update the PR title and description to reflect the changes in this PR. Thank you!

**Comment by sjawhar** [comment]
> LGTM, please update the PR title and description to reflect the changes in this PR. Thank you!

Done

**Comment by dakinggg** [comment]
@sjawhar looks like the tests failed with the recent change

File: llmfoundry/callbacks/eval_output_logging_callback.py [hunk 5]
```diff
@@ -24,51 +25,85 @@ class EvalOutputLogging(Callback):
...
+        **kwargs: Any,
+    ):
         super().__init__(self, *args, **kwargs)
         self.log_tokens = log_tokens
         self.columns = None
         self.name = None
         self.rows = []
+        self.log_output_text = log_output_text
+
+    def init(self, state: State, logger: Logger) -> None:
```
Comment (dakinggg) [inline llmfoundry/callbacks/eval_output_logging_callback.py:42]: to fix the test I think you just need to call `init` with some mock inputs and then it should be fine

Comment (sjawhar) [reply llmfoundry/callbacks/eval_output_logging_callback.py:42]: Done
